### PR TITLE
Restore support for matching a scenario by its tags' and steps' line numbers.

### PR DESCRIPTION
**Pull request by botandrose**
# Description
The Cucumber 4.0 release introduced some regressions in matching scenarios by line numbers other than the exact scenario line. The feature line, tag line, background lines, and step lines all stopped matching their related scenario(s). More context at https://github.com/cucumber/cucumber-ruby/issues/1469

This PR is the first step towards restoring that functionality, and only restores the tags and step lines being matched. It also restores the old tests that used to cover all the matching functionality, pending the ones that still need implementation work.

## Type of change
- Bug fix (non-breaking change which fixes an issue)

# Checklist:
- [x] Tests have been added for any changes to behaviour of the code
- [x] New and existing tests are passing locally and on CI
- [x] `bundle exec rubocop` reports no offenses
- [x] RDoc comments have been updated
- [x] CHANGELOG.md has been updated

### CODE REVIEW

**Comment by botandrose** [comment]
Okay, this PR matches tags as well. It was just too simple to add.

Gonna leave this PR as-is for now, because I'm expecting to do a followup PR to match multiline args (tables and docstrings) again... this is a medium amount of work, but still pretty localized.

Comments, feature declaration lines, and background steps appear to require modification to the pickles compiler, so I'm going to save that for last.

**Comment by luke-hill** [comment]
Where are we at with this PR, would you like it reviewed / considered for merge?

Do you think a sibling PR that adds some more rigorous test cases in cucumber-ruby would be good, so we don't regress again. Whereby we run a scenario from each of the fixed items here (tags / feature keyword / steps).

Happy to help where appropriate or defer items. Just wanting to get the lay of the land so to speak

**Comment by botandrose** [comment]
@luke-hill hi! Yes, sorry, I see that what I said was rather ambiguous! I would indeed like this PR to be reviewed for merge. It fixes the majority of the regression with a simple change, and it's very localized.

Once this is merged, I plan to follow up with more PRs, here and in other repos to fill in the rest of the gaps. And yes, that will incude one in cucumber-ruby verifying the regression is fixed from the outside. I already have a branch for that pushed, even, so I'm already partly there.

So yeah, take a look and let me know what you think of this PR! Thank you for your time.

File: CHANGELOG.md
Comment (luke-hill) [inline CHANGELOG.md:15]: a scenario by tag and step line numbers

File: lib/cucumber/core/test/case.rb
Comment (luke-hill) [inline lib/cucumber/core/test/case.rb:64]: For my benefit as the naming is ambiguous, where is this defined? Github gives me a few opts.

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:123]: it 'matches a location on the first step of the scenario'

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:137]: I don't like this personally because this is technically ambiguous. Comments in the gherkin / pickle e.t.c. are attributed to a line, but not directly attached to a scenario. Whereas a tag **is** directly attached to a gherkin keyword.

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:151]: it 'does not return a matched location on a whitespace line'

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:158]: This scenario is technically a duplicate of the one above for the reason mentioned.

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:177]: in the body of the docstring

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:191]: Scenario can be removed

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (luke-hill) [inline spec/cucumber/core/test/filters/locations_filter_spec.rb:202]: on the table heading

Can also add one for table row(s)

**Review by luke-hill — COMMENTED** [review]
Didn't review the spec from 273 downwards as there's likely some duplicate comments.

Great stuff though.

One question about `xit` - We don't have this standard here. I don't foresee an issue too much if the turnaround time is reasonably quick. But it might be a case where we just don't commit these tests.

File: lib/cucumber/core/test/case.rb
Comment (botandrose) [reply lib/cucumber/core/test/case.rb:64]: Its just a value attr, and it's set here: https://github.com/cucumber/cucumber-ruby-core/blob/baf204008e98900a04a1fa10dbabcc5a90c22c04/lib/cucumber/core/compiler.rb#L42

In more detail, `C::C::Test::Case#location` is a `Location::Precise` value of the Scenario line, set by `C::C::Compiler` when it creates the Test Case upon receiving the relevant information from the pickles compiler. Does this answer your question?

File: spec/cucumber/core/test/filters/locations_filter_spec.rb
Comment (botandrose) [reply spec/cucumber/core/test/filters/locations_filter_spec.rb:137]: I'm with you on this one. I was honestly surprised to see this behavior was ever a part of cucumber. Since it doesn't pass anyways, I'll rm it from this PR, along with the other pending scenarios.

**Comment by botandrose** [comment]
@luke-hill Thank you for the review! I just pushed another commit with the requested changes made. I also removed all the pending specs, to be reintroduced in follow-up PRs when they are passing.

What do you think? Ready to merge?

**Comment by botandrose** [comment]
As I was working on some of the follow-up work to this PR, I realized that this PR also gets us background step matching for free, so I just pushed another commit that adds test coverage.

**Review by luke-hill — APPROVED** [review]

**Comment by luke-hill** [comment]
Next steps (For my benefit).

Merge this
Release this
Update cucumber ruby (Add new test cases in cucumber ruby to prevent regression)
Merge cucumber ruby update
Release new cucumber ruby

**Comment by botandrose** [comment]
Hi Luke, thanks for taking a look and approving the PR! Your next steps look great to me, and I'd like to insert an additional step:

Merge this
-->Merge in an additional PR covering matching locations on multiline arguments<--
Release this
Update cucumber ruby (Add new test cases in cucumber ruby to prevent regression)
Merge cucumber ruby update
Release new cucumber ruby

This additional multiline matching PR is ready to be submitted for review, but I was waiting on this one here to get merged, since the new PR builds on top of it.

**Comment by aslakhellesoy** [comment]
Hi @botandrose,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Pull request merged**
